server/src/main/java/org/opensearch/search/startree/OriginalOrStarTreeQuery.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.search.startree;
+
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.QueryVisitor;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.util.Accountable;
+
+import java.io.IOException;
+
+/**
+ * Preserves star-tree queries which can be used along with original query
+ * Decides which star-tree query to use (or not) based on cost factors
+ *
+ * @opensearch.experimental
+ */
+public class OriginalOrStarTreeQuery extends Query implements Accountable {
+
+    private final StarTreeQuery starTreeQuery;
+    private final Query originalQuery;
+    private boolean starTreeQueryUsed;
+
+    public OriginalOrStarTreeQuery(StarTreeQuery starTreeQuery, Query originalQuery) {
+        this.starTreeQuery = starTreeQuery;
+        this.originalQuery = originalQuery;
+        this.starTreeQueryUsed = false;
+    }
+
+    @Override
+    public String toString(String s) {
+        return "";
+    }
+
+    @Override
+    public void visit(QueryVisitor queryVisitor) {
+
+    }
```

Comment:
We can leverage implementation similar to `IndexOrDocValuesQuery` for these methods